File: C++/total-distance-to-type-a-string-using-one-finger.cpp
// Time:  O(n + 26)
// Space: O(26)

// hash table
const auto& init = []() {
    static const vector<string> KEYBOARD = {"qwertyuiop", "asdfghjkl", "zxcvbnm"};
    vector<pair<int, int>> lookup(26);
    for (int i = 0; i < size(KEYBOARD); ++i) {
        for (int j = 0; j < size(KEYBOARD[i]); ++j) {
            lookup[KEYBOARD[i][j] - 'a'] = {i, j};
        }
    }
    return lookup;
};

const auto& LOOKUP = init();
class Solution {
public:
    int totalDistance(string s) {
        int result = 0;
        pair<int, int> prev = LOOKUP[0];
        for (const auto& x : s) {
            auto curr = LOOKUP[x - 'a'];
            result += abs(curr.first - prev.first) + abs(curr.second - prev.second);
            prev = curr;
        }
        return result;
    }
};

File: Python/total-distance-to-type-a-string-using-one-finger.py
# Time:  O(n + 26)
# Space: O(26)

# hash table
KEYBOARD = ("qwertyuiop", "asdfghjkl", "zxcvbnm")
LOOKUP = [None]*26
for r, row in enumerate(KEYBOARD):
    for c, x in enumerate(row):
        LOOKUP[ord(x)-ord('a')] = (r, c)
class Solution(object):
    def totalDistance(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        prev = LOOKUP[0]
        for x in s:
            curr = LOOKUP[ord(x)-ord('a')]
            result += abs(curr[0]-prev[0])+abs(curr[1]-prev[1])
            prev = curr
        return result
